lib/SPIRVProducerPass.cpp

```diff
@@ -496,8 +427,40 @@ struct SPIRVProducerPass final : public ModulePass {
   // Returns true if |Arg| is called with a coherent resource.
   bool CalledWithCoherentResource(Argument &Arg);
 
+  //
+  // Primary interface for adding SPIRVInstructions to a SPIRVSection.
+  template <enum SPIRVSection TSection = kFunctions>
+  SPIRVID addSPIRVInst(spv::Op Opcode, SPIRVOperandVec &Operands) {
+    bool has_result, has_result_type;
+    spv::HasResultAndType(Opcode, &has_result, &has_result_type);
+    SPIRVID ID = has_result ? incrNextID() : 0;
+    SPIRVInstruction *I = new SPIRVInstruction(Opcode, ID, Operands);
+    SPIRVSections[TSection].push_back(I);
+    return ID;
+  }
+  template <enum SPIRVSection TSection = kFunctions>
+  SPIRVID addSPIRVInst(spv::Op Op) {
+    SPIRVOperandVec Ops;
+    return addSPIRVInst<TSection>(Op, Ops);
+  }
+  template <enum SPIRVSection TSection = kFunctions>
+  SPIRVID addSPIRVInst(spv::Op Op, uint32_t V) {
+    SPIRVOperandVec Ops;
+    Ops << MkNum(V);
+    return addSPIRVInst<TSection>(Op, Ops);
+  }
+  template <enum SPIRVSection TSection = kFunctions>
+  SPIRVID addSPIRVInst(spv::Op Op, const char *V) {
+    SPIRVOperandVec Ops;
+    Ops << MkString(V);
+    return addSPIRVInst<TSection>(Op, Ops);
+  }
+
 private:
   static char ID;
+
+  Module *module;
```

Comment:
I dislike the mixing this introduces. It's ok to store the module pointer, but I'd rather stop passing around the module when that happens. This would ideally be changed in a separate PR.

Reply:
It was a convenience in a few places but I will forgo this for now and make another PR.

Reply:
Actually needed in getSPIRVType unless I capture the DataLayout.

Reply:
For clarity, I was proposing a first PR to make that change, then this one.

Reply:
Working on it..